File: python/hard/trappingrainwater.py
class Solution:
    def trap(self, height: List[int]) -> int:
        stack = []
        ret = 0

        n = len(height)
        for i in range(n):
            while stack and height[stack[-1]] < height[i]:
                pHeight = height[stack[-1]]
                stack.pop()

                if not stack:
                    break

                dist = i - stack[-1] - 1
                minHeight = min(height[stack[-1]], height[i]) - pHeight
                ret += dist * minHeight

            stack.append(i)

        return ret

File: c++/hard/trappingrainwater.cpp
#include <iostream>

class Solution {
public:
    int trap(vector<int>& height) {
        stack<int> stack;
        int ans = 0;

        for(int i = 0; i < height.size(); i++) {
            while(!stack.empty() && height[stack.top()] < height[i]) {
                int popHeight = height[stack.top()];
                stack.pop();

                if(stack.empty()) break;

                int dist = i - stack.top() - 1;
                int minHeight = min(height[stack.top()], height[i]) - popHeight;
                ans += dist * minHeight;
            }
            stack.push(i);
        }

        return ans;
    }
};